core/state_processor.go

```diff
@@ -683,12 +689,75 @@ func (p *StateProcessor) Process(block *types.WorkObject, batch ethdb.Batch) (ty
 	}
 	// Add an etx for each workshare for it to be rewarded
 	for _, uncle := range block.Uncles() {
-		reward := misc.CalculateReward(uncle)
+		reward := misc.CalculateReward(parent, uncle)
 		uncleCoinbase := uncle.PrimaryCoinbase()
-
 		emittedEtxs = append(emittedEtxs, types.NewTx(&types.ExternalTx{To: &uncleCoinbase, Gas: params.TxGas, Value: reward, EtxType: types.CoinbaseType, OriginatingTxHash: origin, ETXIndex: uint16(len(emittedEtxs)), Sender: uncleCoinbase, Data: []byte{uncle.Lock()}}))
 	}
 
+	updatedTokenChoiceSet, err := CalculateTokenChoicesSet(p.hc, parent, emittedEtxs)
+	if err != nil {
+		return nil, nil, nil, nil, 0, 0, 0, nil, err
+	}
+	var exchangeRate *big.Int
+	var beta0, beta1 *big.Float
+	if parent.NumberU64(common.ZONE_CTX) > params.ControllerKickInBlock {
+		exchangeRate, beta0, beta1, err = CalculateExchangeRate(p.hc, parent, updatedTokenChoiceSet)
+		if err != nil {
+			return nil, nil, nil, nil, 0, 0, 0, nil, err
+		}
+	} else {
+		exchangeRate = parent.ExchangeRate()
+		betas := rawdb.ReadBetas(p.hc.headerDb, parent.Hash())
+		beta0 = betas.Beta0()
+		beta1 = betas.Beta1()
+	}
+	err = rawdb.WriteTokenChoicesSet(batch, block.Hash(), &updatedTokenChoiceSet)
+	if err != nil {
+		return nil, nil, nil, nil, 0, 0, 0, nil, err
+	}
+	err = rawdb.WriteBetas(batch, block.Hash(), beta0, beta1)
+	if err != nil {
+		return nil, nil, nil, nil, 0, 0, 0, nil, err
+	}
+	if block.ExchangeRate().Cmp(exchangeRate) != 0 {
+		return nil, nil, nil, nil, 0, 0, 0, nil, fmt.Errorf("invalid exchange rate used (remote: %d local: %d)", block.ExchangeRate(), exchangeRate)
+	}
+	for _, etx := range emittedEtxs {
+		// If the etx is conversion
+		if types.IsConversionTx(etx) {
+			value := etx.Value()
+			// If sender is in Quai, convert the value into Qi
+			if etx.ETXSender().IsInQuaiLedgerScope() {
+				value = misc.QuaiToQi(block, value)
+			}
+			// If sender is in Qi, convert the value into Quai
+			if etx.ETXSender().IsInQiLedgerScope() {
```

Comment:
I don't think Qi has an ETXSender. Check the to address for this

Reply:
i agree, changed it to checks